contrib/test.bzl

```diff
@@ -0,0 +1,126 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A wrapper around container structure tests for Bazel.
+
+This rule feeds a built image and a set of config files
+to the container structure test framework."
+"""
+
+load(
+    "//container:bundle.bzl", "container_bundle",
+)
+
+def _impl(ctx):
+    config_str = ' '.join(['$(pwd)/' + c.short_path for c in ctx.files.configs])
+
+    if ctx.attr.driver == 'tar':
+        # no need to load if we're using raw tar
+        load_statement = ''
+        image_name = '$(pwd)/' + ctx.file.image_tar.short_path
+
+    else:
+        # Since we're always bundling/renaming the image in the macro, this is valid.
+        load_statement = 'docker load -i %s' % ctx.file.image_tar.short_path
+        image_name = ctx.attr.image_name
+
+    # Generate a shell script to execute structure_tests with the correct flags.
+    ctx.actions.expand_template(
+        template=ctx.file._structure_test_tpl,
+        output=ctx.outputs.executable,
+        substitutions={
+          "%{load_statement}": load_statement,
+          "%{configs}": config_str,
+          "%{workspace_name}": ctx.workspace_name,
+          "%{test_executable}": ctx.executable._structure_test.short_path,
+          "%{image}": image_name,
+          "%{driver}": ctx.attr.driver,
+        },
+        is_executable=True
+    )
+
+    return struct(runfiles=ctx.runfiles(files = [
+            ctx.executable._structure_test,
+            ctx.executable.image_tar,
+            ctx.file.image_tar] +
```

Comment:
Can you put the close-bracket on its own line?  I think it will make it clearer that you switch from `+`ing individual files vs. lists

Reply:
done